man/man8/tipc.8
.TH TIPC 8 "02 Jun 2015" "iproute2" "Linux"
.SH NAME
tipc \- a TIPC configuration and management tool
.SH SYNOPSIS

.ad l
.in +8
.ti -8
.B tipc
.RI "[ " OPTIONS " ] " COMMAND " " ARGUMENTS "
.sp

.ti -8
.IR COMMAND " := { "
.BR bearer " | " link " | " media " | " nametable " | " node " | " socket " }
.sp

.ti -8
.IR OPTIONS " := { "
\fB\-h\fR[\fIhelp\fR] }

.SH DESCRIPTION
The Transparent Inter-Process Communication (TIPC) protocol offers total address
transparency between processes which allows applications in a clustered computer
environment to communicate quickly and reliably with each other, regardless of
their location within the cluster.

TIPC originated at the telecommunications manufacturer Ericsson. The first open
source version of TIPC was created in 2000 when Ericsson released its first
Linux version of TIPC. TIPC was introduced in the mainline Linux kernel in 2006
and is now widely used both within and outside of Ericsson.

.SH OPTIONS

.TP
.BR "\-h" , " --help"
Show help about last given command. For example
.B tipc bearer --help
will show bearer help and
.B tipc --help
will show general help. The position of the option in the string is irrelevant.

.TP
.BR "\-j", " \-json"
Output results in JavaScript Object Notation (JSON).

.TP
.BR "\-p", " \-pretty"
The default JSON format is compact and more efficient to parse but hard for most users to read.
This flag adds indentation for readability.

.SH COMMANDS

.TP
.B BEARER
- Show or modify TIPC bearers

.TP
.B LINK
- Show or modify TIPC links

.TP
.B MEDIA
- Show or modify TIPC media

.TP
.B NAMETABLE
- Show TIPC nametable

.TP
.B NODE
- Show or modify TIPC node parameters

.TP
.B SOCKET
- Show TIPC sockets

.SH ARGUMENTS

Command arguments are described in a command specific man page and typically
consist of nested commands along with key value pairs.
If no arguments are given a command typically shows its help text. The explicit
help option
.B -h
or
.B --help
can occur anywhere among the arguments and will show help for the last valid
command given.

.SH EXIT STATUS
Exit status is 0 if command was successful or a positive integer upon failure.

.SH SEE ALSO
.BR tipc-bearer (8),
.BR tipc-link (8),
.BR tipc-media (8),
.BR tipc-nametable (8),
.BR tipc-node (8),
.BR tipc-peer (8),
.BR tipc-socket (8)
.br
.SH REPORTING BUGS
Report any bugs to the Network Developers mailing list
.B <netdev@vger.kernel.org>
where the development and maintenance is primarily done.
You do not have to be subscribed to the list to send a message there.

.SH AUTHOR
Richard Alpe <richard.alpe@ericsson.com>


ChangeLog
2006-03-21  James Lentini <jlentini@netapp.com>
	
	* Increase size of hw address allowed for ip neigh to allow
	  for IB.

2006-03-14  Russell Stuart <russell-lartc@stuart.id.au>
	
	* Fix missing memset in tc sample
	* Fixes for tc hash samples
	* Add sample divisor

2006-03-10  Alpt <alpt@freaknet.org>

	* Add more rt_proto values

2006-03-10 Dale Sedivec <darkness@caliginous.net>
	
	* Warn when using "handle" instead of "classid" with "tc class"

2006-03-10  Jean Tourrilhes <jt@hpl.hp.com>

	* Fix endless loop in netlink error handling


2006-03-10  Stephen Hemminger  <shemminger@osdl.org>

	* Change default lnstat count to 1
	* Update to 2.6.16 headers
	* Add fake version of include/linux/socket.h to fix warnings

2006-01-12  Patrick McHardy <kaber@trash.net>

	* Handle DCCP in ipxfrm.c to allow using port numbers in the selector.

2006-01-10  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Add ip link ntable

2006-01-10  Stephen Hemminger  <shemminger@osdl.org>

	* Update headers to santized kernel 2.6.15
	* Fix ipv6 priority option in u32 
	
2006-01-03 Alpt <alpt@freaknet.org>

	* Ip man page addition

2006-01-03  Jamal Hadi Salim <hadi@znyx.com>

	* Documentation for ifb

2005-12-09  Stephen Hemminger  <shemminger@osdl.org>

	* Add corrupt feature to netem

2005-12-02  Stephen Hemminger  <shemminger@osdl.org>

	* Backout ambigious ip command matches

2005-11-22  Stephen Hemminger  <shemminger@osdl.org>

	* Handle ambigious ip command matches

2005-11-22  Patrick McHardy <kaber@trash.net>

	* Add back ip command aliases

2005-11-07  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Updating for 2.6.14
	- Show UPD{SA,POLICY} message information from kernel instead of error
	- Add lengh check of deleting message from kernel
	- Use macro for struct xfrm_user{sa,policy}_id

	* Minor fix:
	- Add fflush at the end of normal dump

2005-11-01  Jamal Hadi Salim <hadi@znyx.com>

	* Fix handling of XFRM monitor and state

2005-11-01  Stephen Hemminger  <shemminger@osdl.org

	* Update to 2.6.14 sanitized headers

2005-10-24  Patrick McHardy <kaber@trash.net>

	* Fix ip commnad shortcuts

2005-10-12  Stephen Hemminger  <shemminger@osdl.org>

	* Add more CBQ examples from Fedora Core
	* Fix buffer overrun in iproute because of bits vs. bytes confusion
	
2005-10-12  Jamal Hadi Salim <hadi@znyx.com>

	* Fix ip rule flush, need to reopen rtnl

2005-10-07  Stephen Hemminger  <shemminger@osdl.org>

	* Reenable ip mroute

2005-10-07  Mike Frysinger <vapier@gentoo.org>

	* Handle pfifo_fast that has no qopt without segfaulting

2005-10-05  Mads Martin Joergensen <mmj@suse.de>

	* Trivial netem ccopts

2005-10-04  Jerome Borsboom <j.borsboom@erasmusmc.nl>

	* Fix regression in ip addr (libnetlink) handling

2005-09-21  Stephen Hemminger  <shemminger@osdl.org>

	* Fix uninitialized memory and leaks with valgrind
	  Reported by Redhat

2005-09-01   Mike Frysinger <vapier@gentoo.org>

	* Fix build issues with netem tables (parallel make and HOSTCC)
	
2005-09-01  Stephen Hemminger  <shemminger@osdl.org>

	* Integrate support for DCCP into 'ss' (from acme)
	* Add -batch option to ip.
	* Update to 2.6.14 headers

2005-09-01  Eric Dumazet <dada1@cosmosbay.com>
	
	* Fix lnstat : First column should not be summed

2005-08-16  Stephen Hemminger  <shemminger@osdl.org>

	* Limit ip route flush to 10 rounds.
	* Cleanup ip rule flush error message
	
2005-08-08  Stephen Hemminger  <shemminger@osdl.org>

	* Update to 2.6.13+ kernel headers
	* Fix array overrun in paretonormal
	* Fix ematch to not include dropped fields from skb.
	
2005-07-14  Thomas Graf <tgraf@suug.ch>

	* Make ematch bison/lex build with common flex
	
2005-07-10  Stephen Hemminger  <shemminger@osdl.org>
	
	* Fix Gcc 4.0 build warnings signed/unsigned

2005-06-23  Jamal Hadi Salim <hadi@znyx.com>

	* Fix for options process with ipt

2005-06-23  Thomas Graf <tgraf@suug.ch>
	
	* Add extended matches (nbyte, cmp, u32, meta)
	* Add basic classifier
	* Fix clean/distclean makefile targets
	* update local header file copies
	* IPv4 multipath algorithm selection support
	* cscope Makefile target
	* Fix off-by-one while generating argument vector
	  in batched mode.
	* Assume stdin if no argument is given to -batch

2005-06-22  Stephen Hemminger  <shemminger@osdl.org>

	* Update include files to 2.6.12
	* Add ss support for TCP_CONG

2005-06-13  Steven Whitehouse <steve@chygwyn.com>

	* Decnet doc's update

2005-06-07  Stephen Hemminger  <shemminger@osdl.org>

	* Fix 'ip link' map to handle case where device gets autoloaded
	  by using if_nametoindex as fallback
	* Device indices are unsigned not int.

2005-06-07   Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* [ip] show timestamp when using '-t' option.
	* [ip] remove duplicated code for expired message of xfrm.
	* [ip] add "deleteall" command for xfrm;
	  "flush" uses kernel's flush interface and
	  "deleteall" uses legacy iproute2's flush feature like
	   getting-and-deleting-for-each.

2005-03-30  Stephen Hemminger  <shemminger@osdl.org>

	* include/linux/netfilter_ipv4/ip_tables.h dont include compiler.h
	  because it isn't needed and not on all systems
	* Update rtnetlink.h and pkt_cls.h to be stripped versions
	  of headers from 2.6.12-rc1

2005-03-30  Jamal Hadi Salim <hadi@znyx.com>

	* Proper verison of iptables headers (from 1.3.1)
	* Set revision file in m_ipt
	* Fix action_util naming in mirred
	* don't call ll_init_map in mirred

2005-03-19  Thomas Graf <tgraf@suug.ch>

	* Warn about wildcard deletions and provide IFA_ADDRESS upon
	  deletions to enforce prefix length validation for IPv4.
	* Fix netlink message alignment when the last routing attribute added
	  has a data length not aligned to RTA_ALIGNTO.
	
2005-03-30  Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* ipv6 xfrm allocspi and monitor support.
	
2005-03-29  Stephen Hemminger  <shemminger@osdl.org>

	* switch to stack for netem tables

2005-03-18  Stephen Hemminger  <shemminger@osdl.org>

	* add -force option to batch mode

tipc/tipc.c
/* SPDX-License-Identifier: GPL-2.0-or-later */
/*
 * tipc.	TIPC utility frontend.
 *
 * Authors:	Richard Alpe <richard.alpe@ericsson.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>
#include <linux/tipc_netlink.h>
#include <libmnl/libmnl.h>
#include <errno.h>

#include "mnl_utils.h"
#include "bearer.h"
#include "link.h"
#include "nametable.h"
#include "socket.h"
#include "media.h"
#include "node.h"
#include "peer.h"
#include "cmdl.h"
#include "utils.h"

int help_flag;
int json;
struct mnlu_gen_socket tipc_nlg;

static void about(struct cmdl *cmdl)
{
	fprintf(stderr,
		"Transparent Inter-Process Communication Protocol\n"
		"Usage: %s [OPTIONS] COMMAND [ARGS] ...\n"
		"\n"
		"Options:\n"
		" -h, --help \t\tPrint help for last given command\n"
		" -j, --json \t\tJson format printouts\n"
		" -p, --pretty \t\tpretty print\n"
		"\n"
		"Commands:\n"
		" bearer                - Show or modify bearers\n"
		" link                  - Show or modify links\n"
		" media                 - Show or modify media\n"
		" nametable             - Show nametable\n"
		" node                  - Show or modify node related parameters\n"
		" peer                  - Peer related operations\n"
		" socket                - Show sockets\n",
		cmdl->argv[0]);
}

int main(int argc, char *argv[])
{
	int i;
	int res;
	struct cmdl cmdl;
	const struct cmd cmd = {"tipc", NULL, about};
	struct option long_options[] = {
		{"help", no_argument, 0, 'h'},
		{"json", no_argument, 0, 'j'},
		{"pretty", no_argument, 0, 'p'},
		{0, 0, 0, 0}
	};
	const struct cmd cmds[] = {
		{ "bearer",	cmd_bearer,	cmd_bearer_help},
		{ "link",	cmd_link,	cmd_link_help},
		{ "media",	cmd_media,	cmd_media_help},
		{ "nametable",	cmd_nametable,	cmd_nametable_help},
		{ "node",	cmd_node,	cmd_node_help},
		{ "peer",	cmd_peer,	cmd_peer_help},
		{ "socket",	cmd_socket,	cmd_socket_help},
		{ NULL }
	};

	do {
		int option_index = 0;

		i = getopt_long(argc, argv, "hjp", long_options, &option_index);

		switch (i) {
		case 'h':
			/*
			 * We want the help for the last command, so we flag
			 * here in order to print later.
			 */
			help_flag = 1;
			break;
		case 'j':
			/*
			 * Enable json format printouts
			 */
			json = 1;
			break;
		case 'p':
			/*
			 * Enable json pretty output
			 */
			pretty = 1;
			break;
		case -1:
			/* End of options */
			break;
		default:
			/* Invalid option, error msg is printed by getopts */
			return 1;
		}
	} while (i != -1);

	cmdl.optind = optind;
	cmdl.argc = argc;
	cmdl.argv = argv;

	res = mnlu_gen_socket_open(&tipc_nlg, TIPC_GENL_V2_NAME,
				   TIPC_GENL_V2_VERSION);
	if (res) {
		fprintf(stderr,
			"Unable to get TIPC nl family id (module loaded?)\n");
		return -1;
	}

	res = run_cmd(NULL, &cmd, cmds, &cmdl, &tipc_nlg);
	if (res != 0) {
		mnlu_gen_socket_close(&tipc_nlg);
		return -1;
	}

	mnlu_gen_socket_close(&tipc_nlg);
	return 0;
}
